satellite/satellitedb/accounting.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/accounting"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+//database implements DB
+type accountingDB struct {
+	db *dbx.DB
+}
+
+// LastGranularTime records the greatest last tallied bandwidth agreement time
+func (db *accountingDB) LastGranularTime(ctx context.Context) (time.Time, bool, error) {
+	lastBwTally, err := db.db.Find_Timestamps_Value_By_Name(ctx, dbx.Timestamps_Name("LastBandwidthTally"))
+	if lastBwTally == nil {
+		return time.Time{}, true, err
+	}
+	return lastBwTally.Value, false, err
+}
+
+// SaveGranulars records granular tallies (sums of bw agreement values) to the database
+// and updates the LastGranularTime
+func (db *accountingDB) SaveGranulars(ctx context.Context, logger *zap.Logger, latestBwa time.Time, bwTotals map[string]int64) error {
+	// We use the latest bandwidth agreement value of a batch of records as the start of the next batch
+	// This enables us to not use:
+	// 1) local time (which may deviate from DB time)
+	// 2) absolute time intervals (where in processing time could exceed the interval, causing issues)
+	// 3) per-node latest times (which simply would require a lot more work, albeit more precise)
+	// Any change in these assumptions would result in a change to this function
+	// in particular, we should consider finding the sum of bwagreements using SQL sum() direct against the bwa table
+	if len(bwTotals) == 0 {
+		logger.Warn("In SaveGranulars with empty bwtotals")
+		return nil
+	}
+	//insert all records in a transaction so if we fail, we don't have partial info stored
+	tx, err := db.db.Open(ctx)
+	if err != nil {
+		logger.DPanic("Failed to create DB txn in tally query")
+		return err
+	}
+	defer func() {
+		if err == nil {
```

Comment:
I believe that since err is defined in the main body, it can be assigned to in the defer and will still be returned by the main body (in short, its a closure).

Reply:
take a look at this https://play.golang.org/p/oaBCITX-Kvf I think defer will be called after return statement

Reply:
Interesting, I guess there are some finer details I didn't understand behind "Deferred functions may read and assign to the returning function's named return values."
https://blog.golang.org/defer-panic-and-recover
https://play.golang.org/p/8Zq7fDgoUhM